### ISSUE TITLE
Enhancement: [no-unnecessary-condition] An option to imply noUncheckedIndexedAccess to be true

### ISSUE THREAD

**Issue by Sainan** [typescript-eslint/typescript-eslint#12298]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

> TypeScript's noUncheckedIndexedAccess compiler option [...] is often unwieldy in real-world usage
— https://typescript-eslint.io/rules/no-unnecessary-condition/#possibly-undefined-indexed-access

Furthermore, Array.at obviously does not work/exist on objects/Records.

So, I propose an option that implies noUncheckedIndexedAccess being true for no-unnecessary-condition to fix the false-positives while not actually having to enable such an unwieldy option. :)

### Fail

```TypeScript
N/A
```

### Pass

```TypeScript
N/A
```

**Comment by bradzacher** [comment]
We can't do this. Typescript had no speculative APIs. They types are evaluated as authored and that's it. That means we can't ask TS "what if this option was turned on, what would the type be?"

So there are three options:
1) typecheck twice - once with the default options and once with the option turned on.
    - this is obviously not feasible for performance reasons. This also could lead to strange behaviour because adding `undefined` randomly could cause certain branches to change, causing vastly different types.
1) force the typecheck for the entire lint run to have the option turned on
    - same as above - the butterfly effect.
1) reimplement the option from scratch inside the lint rule.
    - this is not feasible as there's all manner of control flow analysis that needs to be implemented to make this work.

Ultimately there's no way we can implement this.

We've already added some affordances into the rule in terms of attempting to access inline index access of objects and arrays. That is as far as we can reasonably go.
